packages/clangd-contexts/src/clangd-config.ts

```diff
@@ -0,0 +1,102 @@
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ *******************************************************************************/
+
+import * as fs from 'fs';
+import * as yaml from 'js-yaml';
+import { isFile, toConfigPath } from './util/file-util';
+
+/** Description of a project configuration as encoded in a `.clangd` file. */
+export interface ClangdConfig {
+    CompileFlags: CompileFlags;
+    SourceFile: string;
+}
+
+/** Representation of the compile flags property of the `.clangd` configuration. */
+export interface CompileFlags {
+    Add?: string | string[];
+    Remove?: string | string[];
+    CompilationDatabase?: string;
+}
+
+/** The name of the project's clangd configuration file to which we apply a clangd context. */
+export const CONFIG_FILE_NAME = '.clangd';
+
+/**
+ * If a `.clangd` configuration file at the given path exists, load it into a {@link ClangdConfig}.
+ *
+ * @param configPath path to the clangd configuration file or its parent directory.
+ * @returns the loaded {@link ClangdConfig} or `undefined` if the configuration file does not exist
+ */
+export function load(configPath: string): ClangdConfig | undefined {
+    const configurationFile = toConfigPath(configPath);
+    if (!isFile(configurationFile)) {
+        return undefined;
+    }
+    const config = yaml.load(fs.readFileSync(configurationFile, 'utf8'));
+    if (config && typeof config === 'object') {
+        return createClangdConfig(configurationFile, config);
+    }
+    return undefined;
+}
+
+/**
+ * Create a new `ClangdConfig` object.
+ *
+ * @param configurationFile the configuration file from which the configuration was loaded or to which it will be written
+ * @param configurationData the configuration data loaded from the `configurationFile`, if any. This is generically typed
+ *  because it is usually loaded from YAML and may contain data used by clangd that is not managed by this API
+ * @returns a new clangd configuration object
+ */
+function createClangdConfig(configurationFile: string, configurationData?: object): ClangdConfig {
+    const result = { CompileFlags: {}, SourceFile: configurationFile };
+    if (configurationData) {
+        Object.assign(result, configurationData);
+    }
+    return result;
+}
+
+/**
+ * If a `.clangd` configuration file at the given path exists, load it into a {@link ClangdConfig},
+ * or else create a new empty {@link ClangdConfig}. A new empty file is not created; be sure not
+ * to {@link save} an empty configuration using the result.
```

Comment:
Everywhere this function is called, the result is eventually saved. The `save` function logs an error (should it `return` or `throw` at that point?) if no source file is defined, but the `createClangdConfig` function requires a source file to run, so it appears that this function will always return a result that is well formed according to the expectations of the `save` function. What is the risk of saving a file with an empty configuration?

Reply:
In general, the result will have some configuration update written into it, so it won't be empty. But it may effectively lose data if there was a file that just couldn't be read. I'm updating the `load` function to throw, but then we must decide whether this `loadOrCreate` would suppress that exception or propagate it. I lean towards the latter to prevent unexpected data loss due simply to innocently malformed YAML.